### ISSUE TITLE
Flexible sparse data file (folder) format

### ISSUE THREAD

**Issue by draffelt** [MRtrix3/mrtrix3#620]
Following the first ever official cross-institute MRtrix meeting at ISMRM, I'm going to recall and document as best as possible what we sketched out for the new sparse data file format (i.e. for storing fixel data). @rtabbara is happy to take it on coding if it's well specified. I was onto my second pint when we discussed this, so forgive me if I've missed anything! 

All data is to be stored in a folder, in which we have the following:
1.  **Index file** (4D uint64), where the first volume defines the number of elements per voxel, and the second volume defines the offset (index) to the corresponding data for each element.  The file must be called `index.mif` or `index.nii`. Allowing nifti files may enable users to more easily create fixel files from other packages.
2. **Data file**. We have a number of options here, however I think storing a binary file here makes sense in terms of storage, even if less flexible. We could have a simple dump of the data, with a small header to declare the file type and datatype. Alternatively, we could store it as a 1D image (even if stored as a Nx1x1 image). This would make it compatible with `mrcalc` etc, but may be confusing when users try and view the image in `mrview`. Ideally some compromise would be good, where the file format is simple enough for users to create fixel files from other packages, yet would still be able to be loaded as an image by `mrcalc`, `mrthreshold`, `mrmath` etc (effectively as  a "1D image"). It would be nice to have a unique extension for this type of file (maybe *.dat?). 
3. **Directions file**. This is just a special data file required when the sparse data file is a fixel image. The fixel directions are stored as 3 floats and like index should be have a set/expected file name (like `directions`). While this is really just a special type of data file, we should probably distinguish it in some way (extension or file type in the header) from the other sparse data since we don't want users to be able to input this into fixel commands (see example use cases below). 
## Interacting with fixel commands on the command line

I'm envisioning that commands that interact with fixel files will all require a single data file to be specified (and not the index or directions). This means its up to the command to look for the corresponding index and directions file, and throw an error if not found. It also means that if multiple fixel data values are required (e.g. if fixel AFD and spread are required), then these will need to be input as separate arguments on the command line. 
## Example Usage

To get a feel of how this would work, here are a few use cases:

```
fod2fixel fod.mif -afd fixel/afd.dat
fixelthreshold fixel/afd.dat fixel/mask.dat
```

The first command will generate the folder and the fixel index and directions file. In the second command, it will detect the folder, index and directs already exists and allow the output to be put into the same folder. We need some mechanism for checking that the index and directions file correspond to the output, ideally without having load both index and output and check each voxel. Suggestions welcome. For commands that both input and output a fixel image, we could use a unique id in the header (eg like the TSF time stamp). However issues arise when generating fixel images from other data (e.g. if using `fod2fixel` to generate multiple data files, e.g. `-afd` and `-peak`, in a single folder). Also, if we want to allow the index to be a nifti image, then we can't stuff a unique id in the comments. 

Do we allow a user to run a command from within a fixel folder? From the example above:

```
cd fixel
fixelthreshold afd.dat mask.dat
```

I.e. here we need to detect that the argument does not contain a folder, then look for an index/directions file within the folder of the input file. Same with the output. One issue if we allow this is that most users (to begin with) won't know to specify the output folder when generating fixel files, and they will end up just writing the fixel data/directions/index into the current working directory. This might be ok if they only want data from the same subject in a folder, but gets problematic if they want to try and save multiple fixel images from different subjects in the same folder (without `fixelcorrespondence`, see below). Not sure if we should force users to specify a folder or not.

Rob, one issue that we may encounter is that fod2fixel currently can output multiple maps (e.g. `fod2fixel fod.mif -afd fixel/afd.dat -peaks fixel/peaks.dat`). But I'm sure we can deal with that. Worst case we just restrict it to a single output and make the option an argument by specifying the output type. 
## fixelcfestats input and output

This is a special case where the user does not need to specify each input fixel data file. Instead we will allow them to specify a single folder, containing a single index and directions file, with each subject's data as a separate data file. To get files into this format I'm envisioning something like this:

```
foreach * : fixelcorrespondence IN/fixel/afd.dat ../template_fixel/mask.dat ../all_subjects/IN.dat
```

Where the current working directory contains all subject folders (e.g. p01/, p02/, c01/, c02/ etc). The fixelcorrespondence command ([step 11 in the docs](http://mrtrix.readthedocs.io/en/latest/workflows/fixel_based_analysis.html)) assigns the fixel value of fixels in the subject to the corresponding one in the template fixel analysis mask.  

In terms of output users can just specify a single folder, and all the p-value, t-value, effect size etc data is placed in it.
## Loading fixel images in mrview

When visualising fixel images with the vector plot tool, we should allow the users to choose any file within the folder. If the user selects and index or directions file we colour fixels by direction. If they select a data file, then we colour by value. However, the viewer should also parse all other data files in the folder and enable the user to change (with a drop down menu) the way that fixels are coloured, thresholded, length. This would require dynamically loading the data onto the GPU if requested. 

As you can see, we have a few things to think about before deciding it's worth the effort (and this does not yet include the programmer API!). I'm sure more will also pop up along the way.

**Comment by jdtournier** [comment]
#### Index file

I wouldn't restrict the datatype to `uint64`. We could allow any integer-valued files here, provided the range is sufficient to capture all the fixels. So assuming the number of fixels is less than ~4 billion, `uint32` would be sufficient. I do agree thought that `uint64` is the safer option to default to. But we shouldn't reject input fixel index files on that basis. 

#### Data file

I'd rather store as a NxCx1 image (N is number of fixels, C is number of coefs per fixel), this would immediately leverage existing commands at no additional cost. We could reduce confusion for users with a warning if they try to open an image that is NxCx1 when the index file also exists - but I think this is optional. Let them display it, it might actually prove handy for debugging (e.g. just to check that the values are roughly sensible, etc.). In practice, I have a feeling the images won't display anyway, OpenGL won't support a texture of that size (on my system, the max texture size is 16384 - see the OpenGL info dialog in MRView)...

We could add an additional file format/extension, but that feels like over-engineering to me... The existing formats are more than sufficient to do this, all we'd be trying to achieve here would be to minimise confusion for users, when I'm not convinced there's any need to. Changing the file extension is probably not going to reduce confusion for those people who are going to be confused, and it's probably going to make things a touch more annoying for those users who do understand, since they'll have to implement some other format handler in their own code, for example...

One issue that we haven't touched on is that NIfTI1 is not going to support this - image dimensions are stored as `uint16`... We need to be able to store that dimensions as at least `uint32`, if not `uint64` for really big images. So we'd need to default to `.mif`...

#### Directions file

I don't get why you'd want to treat this file differently from any other data file? I appreciate you don't want people to necessarily clobber it, but this goes for any of the data in there, including the index file. I think it's sufficient to label it as `directions.mif`, and look for it along with the index file when required. You mentioned that there were examples of where this matters, but I can't see any...?

#### Handling input/output

For commands that input and output fixels that correspond to the same index, I would restrict output to the same folder. This guarantees that the index and directions are present and as expected. Most of the time the idea will be to create new data files for the same fixel mapping anyway. If you absolutely need to store all of this in a different folder, users can simply copy/paste the whole thing, or just what they need. 

I don't think there would be any issue operating within the fixel folder, since we'd be requesting the dirname of the input to figure out where the index and/or directions files are. That should simply come out as `.`, which should just work fine out of the box. I don't think we can realistically avoid users dumping fixel files in existing folders and mixing things up, unless we're talking about creating new fixel outputs. In this case, the same logic applies as for overwriting new images: you'd need to supply a `-force` flag to output a _new_ fixel image to an existing folder. Otherwise, we'd still have the same safeguards as for regular images in terms of overwriting existing data files, so that should limit the scope for users accidentally overwriting data if writing fixel data to a folder that contains a loads of other images. 

For things like `fixelcorrespondence` where you get different index files for the inputs than for the output, obviously you'd need to enforce that the target folder is distinct, which is what you have in your example. 

There's probably other edge cases that I can't think of right now, where we need to be able to either output a new fixel image _or_ add data to an existing fixel image. The issue there is verifying that the index file is consistent with what you would have written. The simplest solution is to simply not allow this. The other option is to explicitly check at runtime that every fixel index that would have been written is the same as is on-file. Not difficult to do, and probably no more time-consuming than a straight write-out, but would need to be coded up. 

OK, now I need to get on with all this marking I'm supposed to do...

**Comment by draffelt** [comment]
>  So assuming the number of fixels is less than ~4 billion, uint32 would be sufficient. 

Agreed. I was still thinking in terms of the implementation of the current sparse image offset. However uint32 will be plenty large enough. 

> I'd rather store as a NxCx1 image (N is number of fixels, C is number of coefs per fixel)

I'm happy with this. Especially if I can use mrcalc, mrthreshold, and mrmath out of the box. 

> One issue that we haven't touched on is that NIfTI1 is not going to support this

Good point. Maybe we should also require that the index is also mif. Then we can store the number of elements in the header, and check this against the data file image size to make sure they correspond. If users mix and match files, at least must have the correct number of elements. If we don't check for this and they do mix files, they might end up getting seg faults when trying to index data that is out of range and not knowing what is going on. It's very unlikely that any two fixels files from two subjects would have the same total number of fixels.

> I don't get why you'd want to treat this file differently from any other data file?

OK, I'm coming around to this idea too. I originally was concerned about what happens when a user tries to input a directions file as input to something expecting a scalar. For example, fixel2voxel can sum all fixel values per voxel and output a 3D map. In this case I guess we just need to error if C, the number of coeff per fixel, is more than one. 

If we ensure the direction file has a consistent naming then this will be enough to identify it by the command. For example, if a user inputs the directions file to `fixelreorient`, then we could treat it as such, and reorient and output the orientations only.  Unless we supported overwriting the directions file with `fixelreorient fixel/directions.mif fixel/directions.mif`, the user will need to specify a different folder as output (and then manually copy any associated data files into the new folder). If they supply the data file, then we can read the directions in separately and copy both to the new output directory automatically.  

> For commands that input and output fixels that correspond to the same index, I would restrict output to the same folder. 

Looking at the current list of fixel related commands, I don't think there will be any that input and output fixels that correspond to the same index:

- `fixelcalc`, `fixellog` and `fixelthreshold` will go. 
- `fixelcorrespondence` will output a fixel image with the same index and directions as the input template. However we won't want to be outputting all subject files in the same folder as the template.
- `fixelreorient` will be changing the directions. If we restrict the naming of the directions file, then the output will have to be a different folder (unless we allow overwriting as mentioned above).  
- `fixel2tsf`, `fixel2voxel`, `fixel2sh`, and `fixelstats` don't output fixel images.
- `fixelcfestats` will output to a separate results folder.

> unless we're talking about creating new fixel outputs. In this case, the same logic applies as for overwriting new images: you'd need to supply a -force flag to output a new fixel image to an existing folder.

I think when creating new fixel images this would work. For example let's say a user performs `fod2fixel fod.mif -afd fixel/afd.mif`. Then decides they also want the spead: `fod2fixel fod.mif -spread fixel/spread.mif`. I think making them use force to override the existing index and directions should be a warning enough. If for example they do something stupid like change FOD segmentation parameters and get a different set of fixels, then at least the number of fixels will be different in the index file header, and so it can be caught by subsequent fixel commands. 

The same logic applies for combining all subjects into a single fixel folder for input to `fixelcfestats` using `fixelcorrepondence` and `warp2metric`.  Basically `-force` will be required to overwrite the index when processing every subject.

## Programmer API
Since we are proposing the use of standard mif images in a folder, do we need a dedicated Sparse Image class to handle IO, or do we just leave it up to the programmer and provide a few convenience functions? Something like:
```
 Image<float> fixel_values = Image<float>::open(argument[0]);
 Image<uint32> index = Sparse::load_index (argument[0]);   // which also checks the number of fixels
 Image<float> directions = Sparse::load_directions (argument[0]); // which also checks the number of fixels
```
Also when creating a new fixel image, we don't know it advance how many fixel there are. So in a command like `fod2fixel` we will need to first grow fixel data in a dynamic data structure. Then at the end we could provide a convenience function to write the data structure to an image. e.g.:
```
std::vector<float> afd;
std::vector<std::vector<float>> directions;
// foreach fixel:
  // push back data to afd and directions
Sparse::save (argument[2], afd, 1) // where 1 is the C, the number of coeff per fixel. 
Sparse::save ("directions.mif", directions, 3) // where 1 is the C, the number of coeff per fixel. 
```
Obviously fixel commands that write out images with a known number of elements (e.g. `fixelreorient`) can just memory map the required output file straight away using Image::create();

**Comment by jdtournier** [comment]
OK, I think we're all on the same page then. Just a couple of comments:

> I think when creating new fixel images this would work. For example let's say a user performs fod2fixel fod.mif -afd fixel/afd.mif. Then decides they also want the spead: fod2fixel fod.mif -spread fixel/spread.mif. I think making them use force to override the existing index and directions should be a warning enough. If for example they do something stupid like change FOD segmentation parameters and get a different set of fixels, then at least the number of fixels will be different in the index file header, and so it can be caught by subsequent fixel commands.

I'm not sure about this, it seems to be asking users to do something risky (overwrite the index file) with no means for them to verify that this is/was a safe thing to do. Either users will shy away from doing it, not wanting to screw their existing analyses, or they'll get used to adding `-force` everywhere, and eventually end up screwing up their data. I think we should make it possible to do this with on the fly validation. Not saying it has to be done now, but I think otherwise we'll end up getting into trouble.

#### Programmer API

I'd also be in favour of keeping the API as simple as possible, only adding convenience methods where they make sense, and leveraging the existing image-handling API as much as possible. Realistically we only need a few convenience calls to e.g. verify the existence and validity of the index file, find the matching directions file, write the data out with automatic output of index and/or directions where they're missing, and validation otherwise, etc. I can't think of too many operations that will require messing about at lower levels than that...?

**Comment by rtabbara** [comment]
Sorry to hijack the conversation, but can I also check, the `index.mif` stores both the number of fixels and just the *starting* offset for a given data file &mdash; that is, if we're designing the data file to be NxCx1, then the the offset would correspond to the data image's row number for just the first fixel belonging to a given voxel?

To me, this was the only way I could make sense of using a .mif to store the index file. Because the fixel data for a given voxel would be packed contiguously, it makes sense that we would only need to know where to look-up the first fixel's data?

Sorry if this goes without saying, but wasn't immediately clear to me. Otherwise, let me know if there was something else in mind.

**Comment by jdtournier** [comment]
@rtabbara: spot on... :+1:

**Comment by thijsdhollander** [comment]
Ok, just read through the whole thing; it all sounds very sensible to me.  Just a few quick things that came to mind as I read through, so I won't forget them:

1. I agree we definitely have to be careful with the usage of `-force`.  I feel we should stick to the principle that, in any "normal" (whatever that means) processing scenario, `-force` should never be *required*.  If whatever we construct seems to end up requiring the use of `-force` for normal, typical scenarios... then that definitely indicates we're getting something fundamentally wrong in the design of the user interface.... and we should probably figure that out before we dive in too deep.

2. A suggestion for a new command: `fixelinfo` (along the likes of `mrinfo` and `tckinfo`).  I know the format is very open and all (being just a folder that users can inspect), but I reckon this will make life much more convenient and less anxiety-inducing for your average (Joe) user.  More interestingly: the command could dub as a sort of "fixel-check", i.e., it could check whether the index matches the directions, and it all matches all data files.  If something doesn't match, it could try to provide *specific* information on what's wrong.  The latter could prove to be very helpful when users do mess up (I'm still a strong believer of potentially apocalyptic scenarios), and end up on the forum simply saying "something's messed up, but I don't know what".  We could simply instruct them to copy-paste the output of `fixelinfo`, and avoid 42 back-and-forth posts to figure things out. :smile: 

I must say, I'm actually quite happy with the way this is heading.  Gotta be honest though, and admit that I'm still slightly sceptical about the fact that we might empower users to mess up things... but provided that we implement a range of good commands to do various task (*even* if they are simple convenience commands and could also be achieved by simple combinations of copying, pasting and/or renaming files), I'm hopeful that we can avoid that happening *too* frequently... :smiley: :+1:

**Comment by draffelt** [comment]
@jdtournier, @rtabbara is looking into the new format...so far so good. 

It would be good to finalise this before 3.0 release, and before we head to Stanford at the end of the month.  Krzysztof Gorgolewski and Russell Poldrack (with the support of some other big names) are pushing for a standard data structure for image processing (see their recent paper [here](http://www.frontiersin.org/10.3389/conf.fnins.2015.91.00056/event_abstract)), which uses NIfTI.  They already have support from several [software packages](http://bids.neuroimaging.io/), and [websites](https://openfmri.org/data-organization/).

Given they are keen to standardise the pre-processing for DWI and fMRI, my guess is that they will want NIfTI all the way through the processing (bit annoying for us but fair enough). 

If they are keen to support fixel-based analysis, then obviously a sparse file is a non-standard format. However we noticed the NIfTI-2 standard (#19) uses int64 for storing the image dimensions. And therefore there is no reason why we could not use NIfTI-2 for the index file *and* the value files, which would make the sparse format readable by any other package that supports NIfTI-2. It could be nice if eventually this became the standard for handling sparse files in all packages.

If we did implement NIfTI-2, what do you think the best way to precede is? Only save as NIfTI-2 for sparse files, and NIFTI-1 for the rest?

**Comment by jdtournier** [comment]
I guess there's no reason not to use NIfTI-2 in that case. It's a bit annoying to be forced to use NIfTI when it's actually a bit limited, but I guess at least it's a standard... If we do support it, it's going to be difficult to figure out what the default should be for writing. Unfortunately, as far as I can tell, there's nothing about the file name that would provide a hint as to what the user might be expecting. We could add a standard option to force output to NIfTI-2, defaulting to NIfTI-1 otherwise, but my guess is that'll lead to loads of confusion as people omit this and can't figure out what might have gone wrong. 

I think the safest option is to default to NIfTI-1 _unless_ any of the information to be stored won't fit in a NIfTI-1 header. Currently, the `pixdim` fields are the only ones likely to overflow - everything else is either ignored by MRtrix3 (as in most other packages, I reckon), or has been upgraded from `float` to `double` (so there might be a very minor loss of precision, but no catastrophic failure). So basically, all we'd need is a quick check to see whether any of the image sizes are greater than 32767, in which case we switch to NIfTI-2. I reckon that could also be combined with a mechanism for forcing NIfTI-2 output for those users that genuinely want it. That would still allow for the possibility that users can't actually use their output in other non-NIfTI-2 compliant packages, against their expectations - but in this case the output would be garbage anyway - and we can provide a warning on the command-line when this happens.

More generally, there's a few formats that would benefit from being able to have additional options supplied. For example, it would be good to be able to specify which series to load in a DICOM folder, without user interaction - at the moment, that can be handled relatively easily using a construct like 

    $ echo 0 1 | mrinfo DICOM_folder

when you know the answers that you would have typed at the terminal, but it's not great. I've been thinking about the possibility of a syntax to allow format-specific options, in a way that bypasses the normal option handling, leaving the format handlers themselves to parse these options without creating unnecessary cross-dependencies between the regular command-line parsing and individual formats. This would allow something like:

    $ mrinfo DICOM_folder/[study=2,series=11]

or 

    $ mrconvert in.mif out.nii[version=2]

This would be relatively easy to differentiate from the syntax used for [number sequences](http://mrtrix.readthedocs.io/en/latest/getting_started/image_data.html#multi-file-numbered-image-support), since the contents would contains an `=` sign for an option. This also makes use of the fact that this square bracket notation is already used - we don't need to add a different delimiter, which is a good thing since there aren't many of those that don't get interpreted by the shell (incidentally, the [square brackets are actually interpreted by the shell](https://www.gnu.org/software/bash/manual/bashref.html#Pattern-Matching), but are left as-is when no match is found, which is why it works fine _most_ of the time). 

What do you reckon? Worth another issue, maybe...?

**Comment by draffelt** [comment]
Cool. I'll have a think about all this and get back to you on Monday.

In terms of fixel files defaulting to NIftI-2 if over 32767. This would happen nearly all of the time, even at native resolution.

**Comment by jdtournier** [comment]
> In terms of fixel files defaulting to NIftI-2 if over 32767. This would happen nearly all of the time, even at native resolution.

For fixel files, yes. But the point is would almost _never_ happen for regular images, which is the case that would actually produce problems for users.

**Comment by Lestropie** [comment]
Merged in as part of #785.
